### PR TITLE
GOBBLIN-744: Support cancellation of a Helix workflow via a DELETE Spec.

### PR DESCRIPTION
**Pull request by sv2000**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-744

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
This task supports the ability to interrupt and cancel a running job on a Gobblin Helix cluster via a DELETE Spec submitted to the JobConfigurationManager. The DELETE Spec should have "gobblin.cluster.shouldCancelRunningJobOnDelete" set to true for cancelling a running job. The default behavior is to simply delete the corresponding JobSpec from the JobCatalog. 

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
IntegrationJobCancelViaSpecSuite and ClusterIntegrationTest.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by sv2000** [comment]
@htran1 @yukuai518 @arjun4084346 @shirshanka Please review.

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java [hunk 1]
```diff
@@ -82,6 +97,72 @@ public void testJobShouldComplete()
...
+
+    helixManager.connect();
+
+    while (TaskDriver.getWorkflowContext(helixManager, IntegrationJobCancelViaSpecSuite.JOB_ID) == null) {
+      log.warn("Waiting for the job to start...");
+      Thread.sleep(1000L);
+    }
+
+    // Give the job some time to reach writer, where it sleeps
+    Thread.sleep(2000L);
```
Comment (shirshanka) [inline gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java]: any way to avoid sleeping, but instead use something else to detect if the job has reached the writer?

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java [hunk 2]
```diff
@@ -82,6 +97,72 @@ public void testJobShouldComplete()
...
+    ZNRecord record = zkClient.readData(zNodePath);
+    String targetState = record.getSimpleField("TargetState");
+    Assert.assertEquals(targetState, TargetState.START.name());
+
+    //Add a JobSpec with DELETE verb signalling the Helix cluster to cancel the workflow
+    cancelViaSpecSuite.addJobSpec(IntegrationJobCancelViaSpecSuite.JOB_ID, SpecExecutor.Verb.DELETE.name());
+
+    //Give some time for the FsScheduledJobConfigurationManager to pick up the DELETE spec and send
+    // DeleteJobConfigArrivalEvent.
+    Thread.sleep(3000L);
```
Comment (shirshanka) [inline gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java]: Any way to avoid sleeping but waiting on some other observable event to figure out if delete has been processed?

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterConfigurationKeys.java [hunk 3]
```diff
@@ -159,4 +159,7 @@
   public static final String KILL_DUPLICATE_PLANNING_JOB = GOBBLIN_CLUSTER_PREFIX + "kill.duplicate.planningJob";
   public static final boolean DEFAULT_KILL_DUPLICATE_PLANNING_JOB = true;
 
+  public static final String SHOULD_CANCEL_RUNNING_JOB_ON_DELETE = GOBBLIN_CLUSTER_PREFIX + "shouldCancelRunningJobOnDelete";
```
Comment (shirshanka) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterConfigurationKeys.java]: suggestion: drop the prefix "should" from this config. So config could be: cancelRunningJobOnDelete

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterConfigurationKeys.java [hunk 4]
```diff
@@ -159,4 +159,7 @@
...
+
```
Comment (shirshanka) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterConfigurationKeys.java]: Add the default value for this config here (following the standard gobblin pattern)

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java [hunk 5]
```diff
@@ -324,10 +330,17 @@ public void handleUpdateJobConfigArrival(UpdateJobConfigArrivalEvent updateJobAr
   }
 
   @Subscribe
-  public void handleDeleteJobConfigArrival(DeleteJobConfigArrivalEvent deleteJobArrival) {
+  public void handleDeleteJobConfigArrival(DeleteJobConfigArrivalEvent deleteJobArrival) throws InterruptedException {
     LOGGER.info("Received delete for job configuration of job " + deleteJobArrival.getJobName());
     try {
       unscheduleJob(deleteJobArrival.getJobName());
+      Properties jobConfig = deleteJobArrival.getJobConfig();
+      if (PropertiesUtils.getPropAsBoolean(jobConfig, GobblinClusterConfigurationKeys.SHOULD_CANCEL_RUNNING_JOB_ON_DELETE, "false")) {
```
Comment (shirshanka) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java]: Should the default be false or true? 
It could be argued that the expected behavior on a job spec deletion is that running jobs should be cancelled? 

Are you just trying to maintain backward compatibility of this with current jobs?

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterConfigurationKeys.java (hunk 3 above)
Comment (jhsenjaliya) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterConfigurationKeys.java]: +1, same with variable name :) 
btw, should this be `job.cancelR unningOnDelete`, since it's a job property just like `job.schedule` ?

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java [hunk 6]
```diff
@@ -141,6 +143,10 @@ public GobblinHelixJobScheduler(Properties properties,
                                                   metricsWindowSizeInMin);
 
     this.startServicesCompleted = false;
+
+    this.helixJobStopTimeoutSeconds = ConfigUtils.getLong(ConfigUtils.propertiesToConfig(properties),
```
Comment (jhsenjaliya) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java]: probably not required to change at this point but i see this usage of ConfigUtils.propertiesToConfig over an over as if its O(1). may be it needs separate PR for reusing the created config.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java (hunk 5 above)
Comment (jhsenjaliya) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java]: I think it has to be `False` since to update the job config `handleUpdateJobConfigArrival` method deletes and insert as new job config. Or if we want it to be `True` then the method `handleUpdateJobConfigArrival` has to override this config while updating the jobConfig. #my2Cents.

Comment (sv2000) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java]: Yes, it is to maintain backward compatibility.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java (hunk 6 above)
Comment (sv2000) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java]: Its an easy change.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java (hunk 5 above)
Comment (sv2000) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java]: @jhsenjaliya +1 for noticing that. If the default were true, it will impact the job config updates as well.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterConfigurationKeys.java (hunk 3 above)
Comment (sv2000) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterConfigurationKeys.java]: Since this config is specific to cluster jobs, I added the GOBBLIN_CLUSTER_PREFIX to the config. Maybe, gobblin.cluster.job.cancelRunningOnDelete?

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterConfigurationKeys.java (hunk 4 above)
Comment (sv2000) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinClusterConfigurationKeys.java]: Ack.

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java (hunk 1 above)
Comment (sv2000) [reply gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java]: Modified SleepingTask to write an empty file declaring that it is RUNNING. However, the test case will wait only a bounded amount of time for the existence of this file, to avoid the pathological condition where the file cannot be created and the test case keeps polling for the existence of this file forever.

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java (hunk 2 above)
Comment (sv2000) [reply gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java]: Similar to the comment above. Keep polling the znode for a bounded amount of time before giving up and declaring failure.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/FsScheduledJobConfigurationManager.java [hunk 7]
```diff
@@ -53,27 +51,26 @@ public FsScheduledJobConfigurationManager(EventBus eventBus, Config config, Muta
...
-        (List<Pair<SpecExecutor.Verb, Spec>>) this._specConsumer.changedSpecs().get();
+    List<Pair<SpecExecutor.Verb, JobSpec>> jobSpecs =
+        (List<Pair<SpecExecutor.Verb, JobSpec>>) this._specConsumer.changedSpecs().get();
 
-    for (Pair<SpecExecutor.Verb, Spec> entry : jobSpecs) {
-      Spec spec = entry.getValue();
+    for (Pair<SpecExecutor.Verb, JobSpec> entry : jobSpecs) {
+      JobSpec jobSpec = entry.getValue();
       SpecExecutor.Verb verb = entry.getKey();
       if (verb.equals(SpecExecutor.Verb.ADD) || verb.equals(SpecExecutor.Verb.UPDATE)) {
```
Comment (htran1) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/FsScheduledJobConfigurationManager.java]: Should call `postUpdateJobConfigArrival` for the update case.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/FsScheduledJobConfigurationManager.java [hunk 8]
```diff
@@ -53,27 +51,26 @@ public FsScheduledJobConfigurationManager(EventBus eventBus, Config config, Muta
 
   @Override
   protected void fetchJobSpecs() throws ExecutionException, InterruptedException {
-    List<Pair<SpecExecutor.Verb, Spec>> jobSpecs =
-        (List<Pair<SpecExecutor.Verb, Spec>>) this._specConsumer.changedSpecs().get();
+    List<Pair<SpecExecutor.Verb, JobSpec>> jobSpecs =
```
Comment (htran1) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/FsScheduledJobConfigurationManager.java]: Should make sure that the job configuration manager receives the specs in modification time order.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java [hunk 9]
```diff
@@ -324,10 +332,18 @@ public void handleUpdateJobConfigArrival(UpdateJobConfigArrivalEvent updateJobAr
   }
 
   @Subscribe
-  public void handleDeleteJobConfigArrival(DeleteJobConfigArrivalEvent deleteJobArrival) {
+  public void handleDeleteJobConfigArrival(DeleteJobConfigArrivalEvent deleteJobArrival) throws InterruptedException {
     LOGGER.info("Received delete for job configuration of job " + deleteJobArrival.getJobName());
     try {
       unscheduleJob(deleteJobArrival.getJobName());
+      Properties jobConfig = deleteJobArrival.getJobConfig();
```
Comment (arjun4084346) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java]: This should go to a private method cancelJobIfRequired() coupled with a public util method cancelHelixJob

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java [hunk 10]
```diff
@@ -116,7 +119,8 @@ public GobblinHelixJobScheduler(Properties properties,
     this.jobCatalog = jobCatalog;
     this.metricContext = Instrumented.getMetricContext(new org.apache.gobblin.configuration.State(properties), this.getClass());
 
-    int metricsWindowSizeInMin = ConfigUtils.getInt(ConfigUtils.propertiesToConfig(this.properties),
+    Config jobConfig = ConfigUtils.propertiesToConfig(this.properties);
```
Comment (arjun4084346) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java]: Conversion of all the properties to Config looks unnecessary to me.

Comment (shirshanka) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java]: This should be the recommended pattern, so we can use ConfigUtils for the rest of the config gets, without incurring repeated txformations between properties and Config.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SleepingTask.java [hunk 11]
```diff
@@ -17,24 +17,51 @@
...
 import org.apache.gobblin.runtime.TaskContext;
+import org.apache.gobblin.runtime.TaskState;
 import org.apache.gobblin.runtime.task.BaseAbstractTask;
 
 @Slf4j
 public class SleepingTask extends BaseAbstractTask {
+  public static final String TASK_STATE_FILE_KEY = "task.state.file.path";
+
   private final long sleepTime;
+  private FileSystem fs;
```
Comment (shirshanka) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SleepingTask.java]: Do you need Hadoop FileSystem here? If you use java filesystem, you can set it to delete on process shutdown to ensure cleanup happens.

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java [hunk 12]
```diff
@@ -82,6 +99,95 @@ public void testJobShouldComplete()
...
+
+    //Add a new JobSpec to the path monitored by the SpecConsumer
+    cancelViaSpecSuite.addJobSpec(IntegrationJobCancelViaSpecSuite.JOB_ID, SpecExecutor.Verb.ADD.name());
+
+    //Start the cluster
+    cancelViaSpecSuite.startCluster();
+
+    helixManager.connect();
+
+    while (TaskDriver.getWorkflowContext(helixManager, IntegrationJobCancelViaSpecSuite.JOB_ID) == null) {
```
Comment (shirshanka) [inline gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java]: maybe you can use AssertWithBackoff class from org.apache.gobblin.testing

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java [hunk 13]
```diff
@@ -82,6 +99,95 @@ public void testJobShouldComplete()
...
+    ZNRecord record = zkClient.readData(zNodePath);
+    String targetState = record.getSimpleField("TargetState");
+    Assert.assertEquals(targetState, TargetState.START.name());
+
+    //Add a JobSpec with DELETE verb signalling the Helix cluster to cancel the workflow
+    cancelViaSpecSuite.addJobSpec(IntegrationJobCancelViaSpecSuite.JOB_ID, SpecExecutor.Verb.DELETE.name());
+
+    int j = 0;
+    boolean successFlag = false;
+    while (true) {
```
Comment (shirshanka) [inline gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java]: AssertWithBackoff can be used here as well

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/suite/IntegrationJobCancelSuite.java [hunk 14]
```diff
@@ -26,19 +26,21 @@
...
+  public static final String TASK_STATE_FILE = "/tmp/IntegrationJobCancelSuite/taskState/_RUNNING";
 
   @Override
   protected Map<String, Config> overrideJobConfigs(Config rawJobConfig) {
     Config newConfig = ConfigFactory.parseMap(ImmutableMap.of(
         ConfigurationKeys.SOURCE_CLASS_KEY, "org.apache.gobblin.cluster.SleepingCustomTaskSource",
         ConfigurationKeys.JOB_ID_KEY, JOB_ID,
         GobblinClusterConfigurationKeys.HELIX_JOB_TIMEOUT_ENABLED_KEY, Boolean.TRUE,
-        GobblinClusterConfigurationKeys.HELIX_JOB_TIMEOUT_SECONDS, 10L))
+        GobblinClusterConfigurationKeys.HELIX_JOB_TIMEOUT_SECONDS, 10L, SleepingTask.TASK_STATE_FILE_KEY, TASK_STATE_FILE))
```
Comment (shirshanka) [inline gobblin-cluster/src/test/java/org/apache/gobblin/cluster/suite/IntegrationJobCancelSuite.java]: formatting seems off

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java (hunk 9 above)
Comment (sv2000) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobScheduler.java]: Added a method.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/FsScheduledJobConfigurationManager.java (hunk 8 above)
Comment (sv2000) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/FsScheduledJobConfigurationManager.java]: Done.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SleepingTask.java (hunk 11 above)
Comment (sv2000) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SleepingTask.java]: Used Java File API.

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java (hunk 12 above)
Comment (sv2000) [reply gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java]: Done.

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java (hunk 13 above)
Comment (sv2000) [reply gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java]: Done.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SleepingTask.java [hunk 15]
```diff
@@ -17,24 +17,52 @@
 
 package org.apache.gobblin.cluster;
 
+import java.io.File;
+import java.io.IOException;
+
+import com.google.common.io.Files;
+
 import avro.shaded.com.google.common.base.Throwables;
```
Comment (shirshanka) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SleepingTask.java]: spotted a faulty import from before

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java [hunk 16]
```diff
@@ -51,28 +68,30 @@ public void testJobShouldComplete()
...
     helixManager.connect();
 
     TaskDriver taskDriver = new TaskDriver(helixManager);
 
-    while (TaskDriver.getWorkflowContext(helixManager, IntegrationJobCancelSuite.JOB_ID) == null) {
-      log.warn("Waiting for the job to start...");
-      Thread.sleep(1000L);
-    }
+    AssertWithBackoff asserter1 = AssertWithBackoff.create().maxSleepMs(1000).backoffFactor(1);
+    asserter1.assertTrue(isTaskStarted(helixManager, IntegrationJobCancelSuite.JOB_ID),
```
Comment (shirshanka) [inline gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java]: you could chain the entire call without needing the local variable asserter1

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java [hunk 17]
```diff
@@ -82,6 +101,105 @@ public void testJobShouldComplete()
...
+  }
+
+  private Predicate<Void> isTaskStarted(HelixManager helixManager, String jobId) {
+    return input -> TaskDriver.getWorkflowContext(helixManager, jobId) != null;
+  }
+
+  private Predicate<Void> isTaskRunning(String taskStateFileName) {
+    return input -> {
+      try {
+        FileSystem fs = FileSystem.getLocal(new Configuration());
```
Comment (shirshanka) [inline gobblin-cluster/src/test/java/org/apache/gobblin/cluster/ClusterIntegrationTest.java]: can avoid using hadoop fs here?

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/FsSpecConsumer.java [hunk 18]
```diff
@@ -74,6 +78,8 @@ public FsSpecConsumer(Config config) {
       return null;
     }
 
+    Arrays.sort(fileStatuses, Comparator.comparingLong(FileStatus::getModificationTime));
```
Comment (shirshanka) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/FsSpecConsumer.java]: add a comment for why you're doing it and what you're expecting (ascending sort?)

**Comment by htran1** [comment]
+1
